doc/loginflow.md

```diff
@@ -0,0 +1,313 @@
+# Juju CLI login flow
+
+This is documenting what happens when a user does `juju login -u admin` at a
+point when no user is logged in.  It has been done by observing what messages
+are exchanged between client and controller in the websocket connection and what
+http requests the client makes to the controller.
+
+In the following the controller host is `CONTROLLER_HOST` and the admin password
+is `PASSWORD`.
+
+Read this document for an easy intro to macaroons:
+https://github.com/rescrv/libmacaroons/blob/master/README.  For a more
+theoretical approach you can read this:
+https://static.googleusercontent.com/media/research.google.com/en//pubs/archive/41892.pdf.
+
+## 1. Get websocket connection
+Open a websocket connection to `wss://CONTROLLER_HOST:17070/api`
+
+## 2. In websocket: first Login request (fails)
+
+The first thing that happens in the websocket is an attempt to call the
+`Admin.Login` endpoint. 
+
+Request:
+```json
+{
+    "request-id": 1,
+    "type": "Admin",
+    "version": 3,
+    "request": "Login",
+    "params": {
+        "auth-tag": "user-admin",
+        "credentials": "",
+        "nonce": "",
+        "macaroons": null,
+        "bakery-version": 3,
+        "cli-args": "[...]",
+        "user-data": "",
+        "client-version": "2.9.26"
+    }
+}
+```
+
+Because there are no credentials or macaroons in the request, the response
+contains a macaroon with a third party caveat that needs to be discharged to
+authenticate the user. This macaroon is made in
+`apiserver/authentication/user.go`, `UserAuthenticator.authenticateMacaroons()`.
+
+From looking at that code, The condition of the 3rd party caveat is
+`need-declared username admin`
```

Comment:
The condition attached to the caveat is `is-authenticated-user admin`.
The 3rd party caveat signals that upon discharge, we expect to have declared caveats added for the supplied key, which is "username".
We also add a time-before caveat of 24 hours which acts as the expiry time for discharged macaroon once minted.
When the discharge error is raised, the client will prompt for password and 2fa if needed and hits the supplied endpoint (which is the trusted identity provider, in this case the controller itself for local users). That http server will check the username and password and returns a newly minted, short lived, login macaroon with the required `is-authenticated-user` condition set and op set to LoginOp.

See:
`func (h *registerUserHandler) ServeHTTP(w http.ResponseWriter, req *http.Request)` in `apiserver/registration.go`.

The client then resubmits a macaroon with the declared username caveat set and the authenticator this time interprets the macaroon as valid via `ai, err := a.Allow(ctx, identchecker.LoginOp)`. It then looks up the declared username in the juju model to make sure that user exists. Note that we only allow the op to be login.

Reply:
I'm strugging a bit with understanding this.  Looking at the code, the `registerUserHandler` is served at `/register`.  The only place I found requests for this is in the `cmd/juju/controller/register.go`, `registerCommand.secretKeyLogin` which I don't think happens in this flow?

Not sure how to dig in further at this point.